src/PanelWindow.vala

```diff
@@ -136,28 +135,133 @@ public class Wingpanel.PanelWindow : Gtk.Window {
             return;
         }
 
-        var monitor = monitor_number == -1 ? this.screen.get_primary_monitor () : monitor_number;
-        var position_top = monitor_y - panel_displacement;
-        var scale_factor = this.get_scale_factor ();
+        /**
+        * https://specifications.freedesktop.org/wm-spec/wm-spec-1.5.html#NETWMSTRUT
+        * The _NET_WM_STRUCT_PARTICAL specification does not allow to reserve space for arbitrary rectangles
+        * on the screen. Instead it only allows to reserve space at the borders of screen.
+        * As for multi-monitor layouts the wingpanel can be at the within the screen (and not at the border)
+        * this makes it impossible to reserve the correct space for all possible multi-monitor layouts.
+        * Fortunately for up to 3 monitors there is always a possiblity to reserve the right space by also
+        * using the struct-left and struct-right cardinals.
+        */
+
+        var display = get_display ();
+        var n_monitors = display.get_n_monitors ();
+        long struts[12] = { 0 };
+
+        if (n_monitors == 1) {
+            set_struts_from_top (struts);
+        } else {
+            var other_rects = new GLib.List <Gdk.Rectangle?> ();
+            int screen_width = 0;
+
+            for (var i = 0; i < n_monitors; i++) {
+                var other_rect = display.get_monitor (i).get_geometry ();
+                var end_x = other_rect.x + other_rect.width;
+
+                if (end_x + 1 > screen_width) {
+                    screen_width = end_x + 1;
+                }
+
+                other_rects.append (other_rect);
+            }
+
+            if (has_no_other_monitor_above (other_rects)) {
+                set_struts_from_top (struts);
+            } else if (has_no_monitor_to_left (other_rects)) {
+                set_struts_from_left (struts);
+            } else if (has_no_monitor_to_right (other_rects, screen_width)) {
+                set_struts_from_right (struts, screen_width);
+            } else {
+                warning ("Unable to set struts, because Wingpanel is not at the edge of the Gdk.Screen area.");
```

Comment:
> could we address that case here by just dropping all struts when there are displays on that many sides?

@cassidyjames this is what i am doing already. If it no possible to set structs the old strut get replaced by struts of width and height 0 and a warning is raised

Reply:
Perfect!